### PR TITLE
xone: 0.3 -> 0.3-unstable-2024-03-16

### LINKED ISSUE #295708: Build failure: xone
### Steps To Reproduce

1. build xone

### Build log

```

nix log /nix/store/cbcv9vfynssj9bl0mrxr5zggj9c8rr8m-xone-0.3.drv
warning: The interpretation of store paths arguments ending in `.drv` recently changed. If this command is now failing try again with '/nix/store/cbcv9vfynssj9bl0mrxr5zggj9c8rr8m-xone-0.3.drv^*'
@nix { "action": "setPhase", "phase": "unpackPhase" }
Running phase: unpackPhase
unpacking source archive /nix/store/sxvbr4xpsz2w9zd2mxqsdxl3zkcz6wfi-source
source root is /build/source
@nix { "action": "setPhase", "phase": "patchPhase" }
Running phase: patchPhase
applying patch /nix/store/15qds5bbdym41di9g70krvi9d4wjbqyx-kernel-6.3.patch
patching file bus/bus.c
@nix { "action": "setPhase", "phase": "updateAutotoolsGnuConfigScriptsPhase" }
Running phase: updateAutotoolsGnuConfigScriptsPhase
@nix { "action": "setPhase", "phase": "configurePhase" }
Running phase: configurePhase
no configure script, doing nothing
@nix { "action": "setPhase", "phase": "buildPhase" }
Running phase: buildPhase
build flags: SHELL=/nix/store/4vzal97iq3dmrgycj8r0gflrh51p8w1s-bash-5.2p26/bin/bash -C /nix/store/q3k5l3bf08s4wa5im3lj63vivqj4ml4y-linux-6.8-dev/lib/modules/6.8.0-zen1/build M=\$\(sourceRoot\) VERSION=0.3>
make: Entering directory '/nix/store/q3k5l3bf08s4wa5im3lj63vivqj4ml4y-linux-6.8-dev/lib/modules/6.8.0-zen1/build'
  CC [M]  /build/source/transport/wired.o
  LD [M]  /build/source/xone-wired.o
  CC [M]  /build/source/transport/dongle.o
/build/source/transport/dongle.c:970:10: error: 'struct usb_driver' has no member named 'drvwrap'
  970 |         .drvwrap.driver.shutdown = xone_dongle_shutdown,
      |          ^~~~~~~
/build/source/transport/dongle.c:970:36: error: initialization of 'const struct attribute_group **' from incompatible pointer type 'void (*)(struct device *)' [-Werror=incompatible-pointer-types]
  970 |         .drvwrap.driver.shutdown = xone_dongle_shutdown,
      |                                    ^~~~~~~~~~~~~~~~~~~~
/build/source/transport/dongle.c:970:36: note: (near initialization for 'xone_dongle_driver.dev_groups')
cc1: some warnings being treated as errors
make[2]: *** [/nix/store/q3k5l3bf08s4wa5im3lj63vivqj4ml4y-linux-6.8-dev/lib/modules/6.8.0-zen1/source/scripts/Makefile.build:243: /build/source/transport/dongle.o] Error 1
make[1]: *** [/nix/store/q3k5l3bf08s4wa5im3lj63vivqj4ml4y-linux-6.8-dev/lib/modules/6.8.0-zen1/source/Makefile:1919: /build/source] Error 2
make: *** [/nix/store/q3k5l3bf08s4wa5im3lj63vivqj4ml4y-linux-6.8-dev/lib/modules/6.8.0-zen1/source/Makefile:240: __sub-make] Error 2
make: Leaving directory '/nix/store/q3k5l3bf08s4wa5im3lj63vivqj4ml4y-linux-6.8-dev/lib/modules/6.8.0-zen1/build'
lines 1-32/32 (END)

```

### Notify maintainers

@rhysmdnz 

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
output here
```
 - system: `"x86_64-linux"`
 - host os: `Linux 6.7.9-zen1, NixOS, 24.05 (Uakari), 24.05.20240309.3030f18`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.1`
 - channels(root): `"nixos"`
 - nixpkgs: `/nix/store/nra828scc8qs92b9pxra5csqzffb6hpl-source`
---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

### PR DESCRIPTION
**Pull request by NateWright**
## Description of changes

Fixes #295708 

## Things done

- Built on platform(s)
  - [X] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Built package on my machine and tested with a controller in a game.

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by drupol — CHANGES_REQUESTED** [review]
Hi !

Thanks for your contribution.

I invite you to read the contributor guide and format the PR title and commit message accordingly.

Pro tip: have a look at the commits in `nixpkgs` repository to have a concrete idea.

Another cool reading: https://nixos.wiki/wiki/Update_a_package

Let me know if you need some help.

File: pkgs/os-specific/linux/xone/default.nix [hunk 1]
```diff
@@ -2,24 +2,15 @@
...
-  version = "0.3";
+  version = "0.3-unstable-2024-03-16";
 
   src = fetchFromGitHub {
     owner = "medusalix";
     repo = pname;
-    rev = "refs/tags/v${version}";
-    sha256 = "sha256-h+j4xCV9R6hp9trsv1NByh9m0UBafOz42ZuYUjclILE=";
+    rev = "948d2302acdd6333295eaba4da06d96677290ad3";
+    sha256 = "sha256-srAEw1ai5KT0rmVUL3Dut9R2mNb00AAZVCcINikh2sM=";
```
Comment (drupol) [inline pkgs/os-specific/linux/xone/default.nix:11]: Prefer `hash = ...` to `sha256 = ...` Both currently work, but the former will make it easier if the algorithm ever needs to change.

Since there could be different hashing algorithms in use, this make sense.

More info: https://nixos.org/manual/nixpkgs/stable/#fetchurl

> Nixpkgs contributors are currently recommended to use hash.

File: pkgs/os-specific/linux/xone/default.nix [hunk 2]
```diff
@@ -45,4 +36,3 @@ stdenv.mkDerivation rec {
     platforms = platforms.linux;
   };
 }
-
```
Comment (drupol) [inline pkgs/os-specific/linux/xone/default.nix:48]: Unrelated changes, please revert.

File: pkgs/os-specific/linux/xone/default.nix [hunk 3]
```diff
@@ -2,24 +2,15 @@
 
 stdenv.mkDerivation rec {
```
Comment (drupol) [inline pkgs/os-specific/linux/xone/default.nix:3]: I think we should use the `finalAttrs` pattern here.

The `finalAttrs` pattern will let you remove the `rec` keyword ([see its implementation in Nix](https://github.com/NixOS/nix/blob/2d1d81114d72ace89ce08cd3bc93f4eb27a2975d/src/libexpr/nixexpr.cc#L121)).

Why is this pattern preferred to `rec` ?

Let's take this simple code example:

```nix
mkDerivation rec {
   foo = 1;
   bar = foo + 1;
}
```

and then `.overrideAttrs(old: { foo = 2; })`, you'll get `{ foo = 2; bar = 2; }` while with `finalAttrs` pattern, it would work correctly because it's a real fixed point.

Let me share a couple of useful links regarding the `finalAttrs` pattern:

1. History: https://discourse.nixos.org/t/avoid-rec-expresions-in-nixpkgs/8293
2. Documentation: https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes
3. Recent example of implementation: https://github.com/NixOS/nixpkgs/compare/17f96f7b978e61576cfe16136eb418f74fab9952..9e6ea843e473d34d4f379b8b0d8ef0425a06defe

Feel free to reach out if you need some assistance.

**Review by drupol — CHANGES_REQUESTED** [review]
Further comments.

File: pkgs/os-specific/linux/xone/default.nix [hunk 4]
```diff
@@ -1,27 +1,18 @@
...
   pname = "xone";
-  version = "0.3";
+  version = "0.3-unstable-2024-03-16";
 
   src = fetchFromGitHub {
     owner = "medusalix";
-    repo = pname;
-    rev = "refs/tags/v${version}";
-    sha256 = "sha256-h+j4xCV9R6hp9trsv1NByh9m0UBafOz42ZuYUjclILE=";
+    repo = finalAttrs.pname;
```
Comment (drupol) [inline pkgs/os-specific/linux/xone/default.nix:9]: Please don't use this.

`repo = pname;` is nice for DRY but creates a binding that goes too far.

See further information about this here: https://github.com/nix-community/nixpkgs-lint/issues/21

**Review by drupol — CHANGES_REQUESTED** [review]

File: pkgs/os-specific/linux/xone/default.nix (hunk 2 above)
Comment (NateWright) [reply pkgs/os-specific/linux/xone/default.nix:48]: Is there a recommended formatter? I have been using nixpkgs-fmt which adds the extra new line

Comment (drupol) [reply pkgs/os-specific/linux/xone/default.nix:48]: Your previous commit was removing the last line. That's the reason why I asked you to revert the changes.

Comment (NateWright) [reply pkgs/os-specific/linux/xone/default.nix:48]: Sorry about that my mistake

File: pkgs/os-specific/linux/xone/default.nix [hunk 5]
```diff
@@ -1,27 +1,18 @@
...
-    (fetchpatch {
-      name = "kernel-6.3.patch";
-      url = "https://github.com/medusalix/xone/commit/bbf0dcc484c3f5611f4e375da43e0e0ef08f3d18.patch";
-      hash = "sha256-A2OzRRk4XT++rS6k6EIyiPy/LJptvVRUxoP7CIGrPWU=";
-    })
-  ];
-
   setSourceRoot = ''
-    export sourceRoot=$(pwd)/${src.name}
+    export sourceRoot=$(pwd)/${finalAttrs.src.name}
```
Comment (drupol) [inline pkgs/os-specific/linux/xone/default.nix:15]: I'm unsure about this, I don't think we need an export. I will check tomorrow.

Comment (GoogleBot42) [reply pkgs/os-specific/linux/xone/default.nix:15]: Doing a quick search for instances of `setSourceRoot =` I see both methods.  So it indeed appears unnecessary but the export slips in from time to time.

**Comment by XenGi** [comment]
@drupol I really appreciate your comments on this PR, but it's sole purpose was to bump the version of this package because it's currently broken with linux 6.8+. Can we please fix this first and then work on improvements in another PR? I wanna play my controller games again. ;)

**Review by drupol — APPROVED** [review]
Fair enough! Thanks for implementing all the changes, it LGTM !

**Comment by XenGi** [comment]
Thx for merging! Now we just need to wait for it to land in unstable: https://nixpk.gs/pr-tracker.html?pr=296470
